native/src/binaryRoutePlanner.cpp

```diff
@@ -1205,32 +1205,37 @@ vector<SHARED_PTR<RouteSegmentResult>> convertFinalSegmentToResults(RoutingConte
 																	const SHARED_PTR<RouteSegment>& finalSegment) {
 	vector<SHARED_PTR<RouteSegmentResult>> result;
 	if (finalSegment) {
-        // ctx.routingTime += finalSegment.distanceFromStart; // done by convertFinalSegmentToResults Java / runRouting iOS
-        float correctionTime = finalSegment->opposite == nullptr ? 0 :
-                        finalSegment->distanceFromStart - distanceFromStart(finalSegment->opposite) - distanceFromStart(finalSegment->parentRoute);
-        SHARED_PTR<RouteSegment> thisSegment =  finalSegment->opposite == nullptr ? finalSegment : finalSegment->getParentRoute(); // for dijkstra
-        SHARED_PTR<RouteSegment> segment = finalSegment->isReverseWaySearch() ? thisSegment : finalSegment->opposite;
+		// ctx.routingTime += finalSegment.distanceFromStart; // done by convertFinalSegmentToResults Java / runRouting iOS
+		float correctionTime = finalSegment->opposite == nullptr ? 0 :
+						finalSegment->distanceFromStart - distanceFromStart(finalSegment->opposite) - distanceFromStart(finalSegment->parentRoute);
+		SHARED_PTR<RouteSegment> thisSegment =  finalSegment->opposite == nullptr ? finalSegment : finalSegment->getParentRoute(); // for dijkstra
+		SHARED_PTR<RouteSegment> segment = finalSegment->isReverseWaySearch() ? thisSegment : finalSegment->opposite;
+		SHARED_PTR<RouteSegment> prevSegment;
 		while (segment && segment->getRoad()) {
 			auto res = std::make_shared<RouteSegmentResult>(segment->road, segment->getSegmentEnd(),
 															segment->getSegmentStart());
 			float parentRoutingTime =
 				segment->getParentRoute() != nullptr ? segment->getParentRoute()->distanceFromStart : 0;
 			res->routingTime = segment->distanceFromStart - parentRoutingTime + correctionTime;
-            correctionTime = 0;
+			correctionTime = 0;
+			prevSegment = segment;
 			segment = segment->getParentRoute();
+			prevSegment->parentRoute = nullptr;//clear the chain of memory dependencies
```

Comment:
Please log parentRoute.use_count() before =nullptr. Mostly, there are 2-3+ uses. Therefore, the most of parentRoute structure(s) will be kept in memory after this function.